baseplate/server/__init__.py

```diff
@@ -432,23 +445,111 @@ def load_and_run_shell() -> None:
+                    if kind=='input':
+                        # Generate an RFC 5424 compliant syslog format
+                        write(f"<13>1 {{time.strftime('%Y-%m-%dT%H:%M:%S.%fZ', time.gmtime())}} {{os.uname().nodename}} baseplate-shell {{os.getpid()}} {{message_id}} - {{data}}")
+                    elif kind=='output' and self.log_output:
+                        odata = u'\\n'.join([u'#[Out]# %s' % s
+                                        for s in data.splitlines()])
+                        write(u'%s\\n' % odata)
+                    self.logfile.flush()
+            ip.logger.logstop = None
+            ip.logger.log_write = partial(log_write, ip.logger)
+            ip.logger.log_write(data="Start IPython logging\\n", message_id="ISTR")
+            """
+        ]
         ipython_config.TerminalInteractiveShell.banner2 = banner
+        ipython_config.LoggingMagics.quiet = True
         start_ipython(argv=[], user_ns=env, config=ipython_config)
         raise SystemExit
     except ImportError:
-        import code
-
         newbanner = f"Baseplate Interactive Shell\nPython {sys.version}\n\n"
         banner = newbanner + banner
 
-        # import this just for its side-effects (of enabling nice keyboard
-        # movement while editing text)
         try:
             import readline
 
-            del readline
+            readline.set_completer(Completer(env).complete)
+            readline.parse_and_bind("tab: complete")
+
         except ImportError:
             pass
 
-        shell = code.InteractiveConsole(locals=env)
+        shell = LoggedInteractiveConsole(_locals=env, logpath=console_logpath)
         shell.interact(banner)
+
+
+def _get_shell_log_path() -> str:
+    """Determine where to write shell audit logs."""
+    if _is_containerized():
+        # write to PID 1 stdout for log aggregation
+        return "/proc/1/fd/1"
+    # otherwise write to a local file
+    return "/var/log/.shell_history"
+
+
+def _is_containerized() -> bool:
+    """Determine if we're running in a container based on cgroup awareness for various container runtimes."""
+    if os.path.exists("/.dockerenv"):
+        return True
+
+    try:
+        with open("/proc/self/cgroup") as my_cgroups_file:
+            my_cgroups = my_cgroups_file.read()
+
+            for hint in ["kubepods", "docker", "containerd"]:
+                if hint in my_cgroups:
+                    return True
+    except IOError:
+        pass
+
+    return False
+
+
+def _has_PID1_parent() -> bool:
+    """Determine parent PIDs up the tree until PID 1 or 0 is reached, do this natively"""
+    parent_pid = os.getppid()
+    while parent_pid > 1:
+        with open(f"/proc/{parent_pid}/status", "r") as proc_status:
+            for line in proc_status.readlines():
+                if line.startswith("PPid:"):
+                    parent_pid = int((line.replace("PPid:", "")))
+                    break
+    return bool(parent_pid)
```

Comment:
@spladug here's the proposed path using `/proc/PID/status` to walk upwards natively. will merge this next week